### ISSUE TITLE
Method "getChatHistory" always returns only one message.

### ISSUE THREAD

**Issue by esindger** [tdlib/td#740]
Hello,
what is wrong with the request? Why I receive only one message?

```
[ 3][t 4][1571446901.608677149][Td10.cpp:164][#1][!Td][&td_requests]	Receive request 9: getChatHistory {
  chat_id = -1001341016129
  from_message_id = 0
  offset = 0
  limit = 20
  only_local = false
}
[ 3][t 0][1571446901.608801126][Client.cpp:270][&td_requests]	Begin to wait for updates with timeout 10.000000
[ 3][t 4][1571446901.608780146][MessagesManager26.cpp:132][#1][!GetChatHistoryRequest]	Get history in chat -1001341016129 from server message 2147483647 with offset 0 and limit 20, 3 tries left. Last read inbox message is server message 38884, last read outbox message is server message 38884, have_full_history = false
[ 4][t 4][1571446901.608866215][MessagesManager26.cpp:139][#1][!GetChatHistoryRequest]	Iterator points to server message 38884
[ 3][t 4][1571446901.608880997][MessagesManager26.cpp:205][#1][!GetChatHistoryRequest]	Iterator after applying offset points to server message 38884, offset = 0, limit = 20, from_the_end = true
[ 3][t 4][1571446901.608928204][MessagesManager12.cpp:465][#1][!GetChatHistoryRequest]	Preloading older before server message 38884
[ 3][t 4][1571446901.608942270][MessagesManager9.cpp:415][#1][!GetChatHistoryRequest]	Load messages in chat -1001341016129 from server message 38884 with offset = 0 and limit = 50. 3 tries left
[ 3][t 4][1571446901.608961105][MessagesManager27.cpp:214][#1][!GetChatHistoryRequest]	Get history in chat -1001341016129 from server message 38884 with offset -1 and limit 51 from server
[ 4][t 4][1571446901.608975887][utils.cpp:21][#1][!GetChatHistoryRequest]	Create storer for messages_getHistory {
  peer = inputPeerChannel {
    channel_id = 1341016129
    access_hash = -1151926457721191652
  }
  offset_id = 38884
  offset_date = 0
  add_offset = -1
  limit = 51
  max_id = 0
  min_id = 0
  hash = 0
}
[ 3][t 4][1571446901.609009981][NetQuery.h:347][#1][!GetChatHistoryRequest]	[Query:[id:1114112][tl:0xdcbb8260][state:Query]]
[ 3][t 4][1571446901.609186172][Td13.cpp:291][#1][!GetChatHistoryRequest][&net_query]	Send [Query:[id:1114112][tl:0xdcbb8260][state:Query]] to dispatcher
[ 3][t 4][1571446901.609242916][NetQuery.h:239][#1][!GetChatHistoryRequest][&net_query]	[Query:[id:1114112][tl:0xdcbb8260][state:Query]] [debug:Td: send to NetQueryDispatcher]
[ 3][t 4][1571446901.609273195][NetQuery.h:239][#1][!GetChatHistoryRequest][&net_query]	[Query:[id:1114112][tl:0xdcbb8260][state:Query]] [debug:dispatch]
[ 3][t 4][1571446901.609295130][NetQuery.h:239][#1][!GetChatHistoryRequest][&net_query]	[Query:[id:1114112][tl:0xdcbb8260][state:Query]] [debug:sent to main session multi proxy DcId{5}]
[ 3][t 4][1571446901.609314203][MessagesManager26.cpp:232][#1][!GetChatHistoryRequest]	Have 1 messages out of requested exact 20
[ 3][t 4][1571446901.609325171][MessagesManager26.cpp:238][#1][!GetChatHistoryRequest]	Return 1 messages in result to getChatHistory
[ 3][t 4][1571446901.609341145][Td48.cpp:279][#1][!Td][&td_requests]	Sending result for request 9: messages {
  total_count = 1
  messages = vector[1] {
    message {
      id = 40772829184
      sender_user_id = 885767761
      chat_id = -1001341016129
      sending_state = null
      is_outgoing = false
      can_be_edited = false
      can_be_forwarded = false
      can_be_deleted_only_for_self = false
      can_be_deleted_for_all_users = false
      is_channel_post = false
      contains_unread_mention = false
      date = 1571403808
      edit_date = 0
      forward_info = null
      reply_to_message_id = 0
      ttl = 0
      ttl_expires_in = 0.000000
      via_bot_user_id = 0
      author_signature = ""
      views = 0
      media_album_id = 0
      content = messageChatAddMembers {
        member_user_ids = vector[1] {
          885767761
        }
      }
      reply_markup = null
    }
  }
}
```

**Comment by levlam** [comment]
This is expected and described in the method description: `For optimal performance the number of returned messages is chosen by the library`. See https://core.telegram.org/tdlib/getting-started#getting-chat-messages for more details.

**Comment by chenyu1990** [comment]
@esindger how to solve?

**Comment by ivanstepanovftw** [comment]
Solved like this
```
void send_history_query(
        ssize_t chat_id,
        ssize_t from_message_id,
        ssize_t offset,
        ssize_t history_size,
        bool only_local,
        std::function<void(td_api::object_ptr<td::td_api::Object>)> on_messages_received,
        std::function<void()> on_complete) {
    send_query(td_api::make_object<td_api::getChatHistory>(chat_id, from_message_id, offset, history_size, only_local), [this, chat_id, from_message_id, offset, history_size, only_local, on_messages_received = std::move(on_messages_received), on_complete = std::move(on_complete)](td_api::object_ptr<td::td_api::Object> object) {
        if (object->get_id() == td_api::error::ID) {
            on_messages_received(std::move(object));
            return on_complete();
        }
        auto messages = td::move_tl_object_as<td_api::messages>(object);
        ssize_t next_from_message_id = messages->messages_.back()->id_;
        ssize_t remaining_history_size = history_size - messages->messages_.size();
        bool are_messages_empty = messages->messages_.empty();
        on_messages_received(std::move(messages));
        if (remaining_history_size > 0 && !are_messages_empty) {
            send_history_query(chat_id, next_from_message_id, offset, remaining_history_size, only_local, on_messages_received, on_complete);
        } else {
            on_complete();
        }
    });
}
```

**Comment by kriposaf578** [comment]
> Solved like this
> 
> ```
> void send_history_query(
>         ssize_t chat_id,
>         ssize_t from_message_id,
>         ssize_t offset,
>         ssize_t history_size,
>         bool only_local,
>         std::function<void(td_api::object_ptr<td::td_api::Object>)> on_messages_received,
>         std::function<void()> on_complete) {
>     send_query(td_api::make_object<td_api::getChatHistory>(chat_id, from_message_id, offset, history_size, only_local), [this, chat_id, from_message_id, offset, history_size, only_local, on_messages_received = std::move(on_messages_received), on_complete = std::move(on_complete)](td_api::object_ptr<td::td_api::Object> object) {
>         if (object->get_id() == td_api::error::ID) {
>             on_messages_received(std::move(object));
>             return on_complete();
>         }
>         auto messages = td::move_tl_object_as<td_api::messages>(object);
>         ssize_t next_from_message_id = messages->messages_.back()->id_;
>         ssize_t remaining_history_size = history_size - messages->messages_.size();
>         bool are_messages_empty = messages->messages_.empty();
>         on_messages_received(std::move(messages));
>         if (remaining_history_size > 0 && !are_messages_empty) {
>             send_history_query(chat_id, next_from_message_id, offset, remaining_history_size, only_local, on_messages_received, on_complete);
>         } else {
>             on_complete();
>         }
>     });
> }
> ```
on_messages_receive what is this method?

**Comment by ivanstepanovftw** [comment]
> > Solved like this
> > ```
> > void send_history_query(
> >         ssize_t chat_id,
> >         ssize_t from_message_id,
> >         ssize_t offset,
> >         ssize_t history_size,
> >         bool only_local,
> >         std::function<void(td_api::object_ptr<td::td_api::Object>)> on_messages_received,
> >         std::function<void()> on_complete) {
> >     send_query(td_api::make_object<td_api::getChatHistory>(chat_id, from_message_id, offset, history_size, only_local), [this, chat_id, from_message_id, offset, history_size, only_local, on_messages_received = std::move(on_messages_received), on_complete = std::move(on_complete)](td_api::object_ptr<td::td_api::Object> object) {
> >         if (object->get_id() == td_api::error::ID) {
> >             on_messages_received(std::move(object));
> >             return on_complete();
> >         }
> >         auto messages = td::move_tl_object_as<td_api::messages>(object);
> >         ssize_t next_from_message_id = messages->messages_.back()->id_;
> >         ssize_t remaining_history_size = history_size - messages->messages_.size();
> >         bool are_messages_empty = messages->messages_.empty();
> >         on_messages_received(std::move(messages));
> >         if (remaining_history_size > 0 && !are_messages_empty) {
> >             send_history_query(chat_id, next_from_message_id, offset, remaining_history_size, only_local, on_messages_received, on_complete);
> >         } else {
> >             on_complete();
> >         }
> >     });
> > }
> > ```
> 
> on_messages_receive what is this method?

@kriposaf578, This code snippet is a C++ function that retrieves the chat history for a given chat on Telegram (or a similar messaging platform that uses a similar API structure). It is designed to work with the Telegram Database Library (TDLib), which is a tool used by developers to create custom Telegram clients or to integrate Telegram functionalities into their software.

Here's a breakdown of what the `send_history_query` function does and the role of the `on_messages_received` callback:

### Function Parameters
- `chat_id`: The unique identifier of the chat whose history you want to retrieve.
- `from_message_id`: The message identifier from which to start retrieving the history. This allows for paginated retrieval of the chat history.
- `offset`: A numerical offset that indicates how many messages to skip from the start. It's used for pagination.
- `history_size`: The number of messages to retrieve.
- `only_local`: A boolean indicating whether to retrieve messages from the local cache only, without requesting them from the server.
- `on_messages_received`: A callback function that is called when a batch of messages is received. It's used to process or handle the messages.
- `on_complete`: A callback function that is called when the entire process of fetching the chat history is complete.

### Main Logic
The function sends a query to retrieve chat history based on the provided parameters. When it receives a response, it checks if the response is an error. If it is, it calls the `on_messages_received` with the error object and then calls `on_complete` to indicate that the process is finished.

If the response is not an error, it processes the received messages. It updates the `from_message_id` to the ID of the last message in the received batch and calculates the remaining number of messages to retrieve (`remaining_history_size`). It then calls the `on_messages_received` callback with the received messages.

### Recursion for Pagination
If there are more messages to retrieve (i.e., `remaining_history_size > 0`) and the last batch of messages was not empty, the function recursively calls itself with updated parameters to retrieve the next batch of messages. This process repeats until all requested messages have been retrieved or there are no more messages left. Once the process is complete, the `on_complete` callback is called.

### `on_messages_received` Callback
This is a user-defined function passed as a parameter to `send_history_query`. Its main purpose is to handle or process the messages once they are received. This could involve storing the messages in a database, displaying them to the user, or performing some analysis on the messages. The specific implementation of this callback would depend on the needs of the application using this function.

In summary, `send_history_query` is a function for asynchronously fetching chat history in a paginated manner, with `on_messages_received` being a crucial part of its operation, allowing for flexible processing of the fetched messages.

**Comment by ivanstepanovftw** [comment]
Let's create a simplified example that demonstrates how you might implement the `on_messages_received` callback function and use the `send_history_query` function within a hypothetical application.

Imagine we're building a console application that retrieves and prints the chat history of a specific chat to the console. We'll simplify error handling and other aspects for clarity.

### Step 1: Define `on_messages_received` Callback
This function will be called with a batch of messages. Our task here is to print each message's content to the console.

```cpp
void on_messages_received(td_api::object_ptr<td::td_api::messages> messages) {
    for (const auto& message : messages->messages_) {
        std::cout << "Message ID: " << message->id_ << ", Content: ";
        if (auto textMessage = td::move_tl_object_as<td_api::messageText>(message->content_)) {
            std::cout << textMessage->text_->text_ << std::endl;
        } else {
            std::cout << "[Non-text message]" << std::endl;
        }
    }
}
```

### Step 2: Define `on_complete` Callback
This function simply indicates that the chat history retrieval process is complete.

```cpp
void on_complete() {
    std::cout << "Finished retrieving chat history." << std::endl;
}
```

### Step 3: Usage of `send_history_query`
Let's assume we have a `ChatHistoryRetriever` class that has the `send_history_query` function. We'll create an instance of this class and call the method with our callbacks and desired parameters.

```cpp
int main() {
    // Assume these are initialized and valid.
    ssize_t chat_id = 123456789; // Example chat ID
    ssize_t from_message_id = 0; // Start from the beginning
    ssize_t offset = 0;
    ssize_t history_size = 100; // We want to retrieve 100 messages
    bool only_local = false; // Retrieve messages from server too

    ChatHistoryRetriever retriever;

    // Call the method with our callbacks and parameters.
    retriever.send_history_query(chat_id, from_message_id, offset, history_size, only_local,
                                 on_messages_received, on_complete);

    // The rest of your application logic here...
}
```

In this example, the `send_history_query` function is designed to asynchronously fetch the chat history. The `on_messages_received` callback processes each batch of messages by printing them to the console, and `on_complete` indicates the end of the retrieval process.

Remember, this is a simplified example. In a real-world scenario, you would need to handle errors, manage memory more carefully (especially with raw pointers and asynchronous callbacks), and likely deal with more complex message types and content.
